src/supernova/circuit.rs

```diff
@@ -0,0 +1,754 @@
+//! Supernova implemetation support arbitrary argumented circuits and running instances.
+//! There are two Verification Circuits for each argumented circuit: The primary and the secondary.
+//! Each of them is over a Pasta curve but
+//! only the primary executes the next step of the computation.
+//! Each circuit takes as input 2 hashes.
+//! Each circuit folds the last invocation of the other into the respective running instance, specified by augmented_circuit_index
+//!
+//! The augmented circuit F' for SuperNova that includes everything from Nova
+//!   and additionally checks:
+//!    1. Ui[] are contained in X[0] hash pre-image.
+//!    2. R1CS Instance u is folded into Ui[augmented_circuit_index] correctly; just like Nova IVC.
+//!    3. (optional by F logic) F circuit might check program_counter_{i} invoked current F circuit is legal or not.
+//!    3. F circuit produce program_counter_{i+1} and sent to next round for optionally constraint the next F' argumented circuit.
+
+use crate::{
+  constants::NUM_HASH_BITS,
+  gadgets::{
+    ecc::AllocatedPoint,
+    r1cs::{
+      conditionally_select_alloc_relaxed_r1cs, AllocatedR1CSInstance, AllocatedRelaxedR1CSInstance,
+    },
+    utils::{
+      add_allocated_num, alloc_num_equals, alloc_scalar_as_base, alloc_zero,
+      conditionally_select_vec, le_bits_to_num, scalar_as_base,
+    },
+  },
+  r1cs::{R1CSInstance, RelaxedR1CSInstance},
+  traits::{
+    circuit_supernova::StepCircuit, commitment::CommitmentTrait, Group, ROCircuitTrait,
+    ROConstantsCircuit,
+  },
+  Commitment,
+};
+use bellperson::{
+  gadgets::{
+    boolean::{AllocatedBit, Boolean},
+    num::AllocatedNum,
+    Assignment,
+  },
+  Circuit, ConstraintSystem, SynthesisError,
+};
+use ff::Field;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CircuitParams {
+  limb_width: usize,
+  n_limbs: usize,
+  is_primary_circuit: bool, // A boolean indicating if this is the primary circuit
+}
+
+impl CircuitParams {
+  pub fn new(limb_width: usize, n_limbs: usize, is_primary_circuit: bool) -> Self {
+    Self {
+      limb_width,
+      n_limbs,
+      is_primary_circuit,
+    }
+  }
+
+  pub fn get_n_limbs(&self) -> usize {
+    self.n_limbs
+  }
+}
+
+#[derive(Debug)]
+pub struct CircuitInputs<'a, G: Group> {
+  params_next: G::Scalar,
+  params: G::Scalar,
+  i: G::Base,
+  z0: &'a Vec<G::Base>,
+  zi: Option<&'a Vec<G::Base>>,
+  U: Option<&'a Vec<Option<RelaxedR1CSInstance<G>>>>,
```

Comment:
Those references of vectors are the hallmark of a place where you'd want to use slices (`&[G::Base]`). Running clippy should help you find those opportunities.

Reply:
FYI, the lint you're looking to activate is [`ptr_arg`](https://rust-lang.github.io/rust-clippy/master/#/ptr_arg)

Reply:
Thanks for the hint, applied clippy and fix all warning/errors